npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/pnpm_lockfile_updater.rb

```diff
@@ -0,0 +1,172 @@
+# frozen_string_literal: true
+
+require "dependabot/npm_and_yarn/update_checker/requirements_updater"
+require "dependabot/npm_and_yarn/helpers"
+require "dependabot/npm_and_yarn/update_checker/registry_finder"
+require "dependabot/shared_helpers"
+
+module Dependabot
+  module NpmAndYarn
+    class FileUpdater
+      class PnpmLockfileUpdater
+        require_relative "package_json_updater"
+
+        def initialize(dependencies:, dependency_files:, repo_contents_path:, credentials:)
+          @dependencies = dependencies
+          @dependency_files = dependency_files
+          @repo_contents_path = repo_contents_path
+          @credentials = credentials
+        end
+
+        def updated_pnpm_lock_content(pnpm_lock)
+          @updated_pnpm_lock_content ||= {}
+          return @updated_pnpm_lock_content[pnpm_lock.name] if @updated_pnpm_lock_content[pnpm_lock.name]
+
+          new_content = run_pnpm_update(pnpm_lock: pnpm_lock)
+          @updated_pnpm_lock_content[pnpm_lock.name] = new_content
+        rescue SharedHelpers::HelperSubprocessFailed => e
+          handle_pnpm_lock_updater_error(e, pnpm_lock)
+        end
+
+        private
+
+        attr_reader :dependencies, :dependency_files, :repo_contents_path, :credentials
+
+        IRRESOLVABLE_PACKAGE = "ERR_PNPM_NO_MATCHING_VERSION"
+        INVALID_REQUIREMENT = "ERR_PNPM_SPEC_NOT_SUPPORTED_BY_ANY_RESOLVER"
+        MISSING_PACKAGE = /(?<package_req>.*?) is not in the npm registry, or you have no permission to fetch it/
```

Comment:
That'd be any supported version I think, since I haven't seen any differences between versions. Also, I believe in particular the `ERR_PNPM` codes are there to allow changing the error wording without changing the error code, so those should stay stable. Unfortunately, not all errors have codes like that.

Reply:
Ah, that's brilliant of the `pnpm` folks to ship well-known error codes!